src/transformers/modeling_dpr.py

```diff
@@ -0,0 +1,551 @@
+        assert config.hidden_size > 0, "Encoder hidden_size can't be zero"
+        self.encode_proj = nn.Linear(config.hidden_size, project_dim) if project_dim != 0 else None
+        self.init_weights()
+
+    @classmethod
+    def init_encoder(cls, cfg_name: str, projection_dim: int = 0, dropout: float = 0.0, **kwargs) -> "DprBertEncoder":
+        cfg = BertConfig.from_pretrained(cfg_name)
+        if dropout != 0:
+            cfg.attention_probs_dropout_prob = dropout
+            cfg.hidden_dropout_prob = dropout
+        return cls(cfg, project_dim=projection_dim)
+
+    def forward(self, input_ids: T, token_type_ids: Optional[T], attention_mask: Optional[T]) -> Tuple[T, ...]:
+        if self.config.output_hidden_states:
+            sequence_output, pooled_output, hidden_states = super().forward(
+                input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
+            )
+        else:
+            hidden_states = None
+            sequence_output, pooled_output = super().forward(
+                input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
+            )
+        pooled_output = sequence_output[:, 0, :]
+        if self.encode_proj:
+            pooled_output = self.encode_proj(pooled_output)
+        return sequence_output, pooled_output, hidden_states
+
+    def get_out_size(self) -> int:
+        if self.encode_proj:
+            return self.encode_proj.out_features
+        return self.config.hidden_size
+
+
+###########
+# Reader
+###########
+
+
+class Reader(nn.Module):
+    def __init__(self, encoder: nn.Module, hidden_size):
+        super(Reader, self).__init__()
+        self.encoder = encoder
+        self.qa_outputs = nn.Linear(hidden_size, 2)
+        self.qa_classifier = nn.Linear(hidden_size, 1)
+
+    def forward(self, input_ids: T, attention_mask: T):
+        # notations: N - number of questions in a batch, M - number of passages per questions, L - sequence length
+        N, M, L = input_ids.size()
+        start_logits, end_logits, relevance_logits = self._forward(
+            input_ids.view(N * M, L), attention_mask.view(N * M, L)
+        )
+        return start_logits.view(N, M, L), end_logits.view(N, M, L), relevance_logits.view(N, M)
+
+    def _forward(self, input_ids, attention_mask):
+        sequence_output, _pooled_output, _hidden_states = self.encoder(input_ids, None, attention_mask)
+        logits = self.qa_outputs(sequence_output)
+        start_logits, end_logits = logits.split(1, dim=-1)
+        start_logits = start_logits.squeeze(-1)
+        end_logits = end_logits.squeeze(-1)
+        rank_logits = self.qa_classifier(sequence_output[:, 0, :])
+        return start_logits, end_logits, rank_logits
+
+
+##############
+# Providers
+##############
+
+
+def get_bert_question_encoder_component(config: DprConfig, **kwargs) -> DprBertEncoder:
+    question_encoder = DprBertEncoder.init_encoder(
```

Comment:
Don't really think we need the `init_encoder` function. I'd prefer to have a function that creates the correct config and get rid of the `init_encoder` function.

Reply:
I removed them to use init instead